### PR TITLE
Refactor and v 1.0.0

### PR DESCRIPTION
**Pull request by kmcphillips**
@garethson

Basically a full refactor. Releasing version 1.0.0 after 6 years!

* Extract permalink generation into `String#to_permalink` patch.
* Create a `Config` class to store concerns, and put that in a single class variable rather than class instance variables.
* Use `public_send` over `send`.
* Require at least Ruby 2.0.
* Require at least Rails 4.0.
* Use `ActiveSupport::Concern`.
* Refactor main class macro to be way more readable and use all the above things.

### CODE REVIEW

**Comment by kmcphillips** [comment]
Added way more tests and also fixed some edge case bugs that those bubbled up.

File: lib/config.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+
+      def initialize(options={})
+        @config = options.with_indifferent_access
+
+        @to = @config[:to].try(:to_sym) || :permalink
+        @from = @config[:from].try(:to_sym) || :title
+        @separator = @config[:underscore] ? "_" : "-"
+
+        @max_length = @config[:max_length].to_i rescue 0
+        @max_length = 60 unless @max_length && @max_length > 0
```
Comment (kmcphillips) [inline lib/config.rb]: Arguably these could each be in their own methods.

File: lib/acts_as_permalink.rb [hunk 2]
```diff
@@ -1,69 +1,47 @@
...
-  module Permalink #:nodoc:
+require "string"
+require "config"
 
-    def self.included(base)
-      base.extend ClassMethods
-    end
+module Acts
+  module Permalink
+    extend ActiveSupport::Concern
```
Comment (garethson) [inline lib/acts_as_permalink.rb]: :clap:

File: lib/config.rb (hunk 1 above)
Comment (garethson) [inline lib/config.rb]: I think `@max_length && @max_length > 0` can just be `@max_length > 0`. The preceding line guarantees an integer is present, right?

**Comment by garethson** [comment]
Worthy of a `1.0`! :ship:

File: lib/config.rb (hunk 1 above)
Comment (kmcphillips) [inline lib/config.rb]: I believe you are correct.
